evm/estimators/gas.py

```diff
@@ -0,0 +1,85 @@
+from cytoolz import (
+    curry,
+)
+
+
+from evm.exceptions import (
+    OutOfGas,
+)
+from evm.utils.spoof import (
+    SpoofTransaction,
+)
+
+
+@curry
+def execute_plus_buffer(multiplier, state, transaction):
+    computation = state.execute_transaction(transaction)
+
+    if computation.is_error:
+        raise computation._error
+
+    gas_used = transaction.gas_used_by(computation)
+
+    gas_plus_buffer = int(gas_used * multiplier)
+
+    return min(gas_plus_buffer, state.gas_limit)
+
+
+double_execution_cost = execute_plus_buffer(2)
+
+
+def _is_enough_gas(state, transaction):
+    snapshot = state.snapshot()
+    computation = state.execute_transaction(transaction)
+    state.revert(snapshot)
+
+    if computation.is_error:
+        return not isinstance(computation._error, OutOfGas)
+    else:
+        return True
+
+
+@curry
+def binary_search(tolerance, state, transaction):
```

Comment:
Maybe use slightly more explicit name of something like `binary_search_required_gas` or `binary_gas_search`